google/gax/utils/protobuf.py

```diff
@@ -0,0 +1,168 @@
+# Copyright 2017, Google Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""Utility functions around reading and writing from protobufs."""
+
+import collections
+
+from google.protobuf.message import Message
+
+__all__ = ('get', 'set', 'setdefault')
+
+
+SENTINEL = object()
+
+
+def get(pb_or_dict, key, default=SENTINEL):
+    """Retrieve the given key off of the object.
+
+    If a default is specified, return it if the key is not found, otherwise
+    raise KeyError.
+
+    Args:
+        pb_or_dict: An object which may be either:
```

Comment:
You can basically always send a dictionary to any GAPIC function that requires a `protobuf.message.Message` subclass. This is _huge_ for usability, because it is a lot more palatable to tell a user to issue:

```python
request = {
    'image': {'source': {'image_uri': 'http://bogus.bogus/img.jpg'}},
    'features': ['FACE_DETECTION'],
}
```

...than it is to ask them to send the equivalent protobuf messages (the example above would require _four_ imports).